-- MySQL dump 10.13  Distrib 5.1.35, for Win32 (ia32)
--
-- Host: localhost    Database: yaml
-- ------------------------------------------------------
-- Server version	5.1.35-community

/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;

--
-- Table structure for table `filecontents`
--

DROP TABLE IF EXISTS `filecontents`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `filecontents` (
  `filecontent_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `identifier` varchar(1024) DEFAULT NULL,
  `size` int(255) unsigned NOT NULL,
  PRIMARY KEY (`filecontent_id`) USING BTREE
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `filecontents`
--

LOCK TABLES `filecontents` WRITE;
/*!40000 ALTER TABLE `filecontents` DISABLE KEYS */;
INSERT INTO `filecontents` VALUES (1,'md5fa983d9ef4fbcd64aa33bc8b526aec61',323737);
/*!40000 ALTER TABLE `filecontents` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `filewithname`
--

DROP TABLE IF EXISTS `filewithname`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `filewithname` (
  `filewithname_id` int(10) unsigned NOT NULL,
  `filecontent_id` int(10) unsigned NOT NULL,
  `code_basename` varchar(1024) NOT NULL,
  `directory` varchar(1024) NOT NULL,
  `ctime` datetime NOT NULL,
  `mtime` datetime NOT NULL,
  PRIMARY KEY (`filewithname_id`),
  KEY `filecontent_id` (`filecontent_id`),
  CONSTRAINT `filecontent_id` FOREIGN KEY (`filecontent_id`) REFERENCES `filecontents` (`filecontent_id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `filewithname`
--

LOCK TABLES `filewithname` WRITE;
/*!40000 ALTER TABLE `filewithname` DISABLE KEYS */;
/*!40000 ALTER TABLE `filewithname` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `parameters`
--

DROP TABLE IF EXISTS `parameters`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `parameters` (
  `parameter_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `code_key` varchar(1024) NOT NULL,
  `code_value` varchar(1024) NOT NULL,
  `run_id` int(10) unsigned NOT NULL,
  PRIMARY KEY (`parameter_id`) USING BTREE,
  KEY `run_id` (`run_id`),
  CONSTRAINT `run_id3` FOREIGN KEY (`run_id`) REFERENCES `runs` (`run_id`)
) ENGINE=InnoDB AUTO_INCREMENT=494 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `parameters`
--

LOCK TABLES `parameters` WRITE;
/*!40000 ALTER TABLE `parameters` DISABLE KEYS */;
INSERT INTO `parameters` VALUES (1,'Write identified proteins by spectral count','yes',4),(2,'Write identified proteins by spot','yes',4),(3,'Write identified proteins by spectral count','yes',5),(4,'Write identified proteins by spot','yes',5),(5,'Write identified proteins by distinct peptide count','yes',5),(6,'Write additional peptides identified by GPF','yes',5),(7,'Write ambiguous peptides','yes',5),(8,'Write modified peptides','yes',5),(9,'Putative entry prefix','__putative__',5),(10,'Output directory','X:/AK Hippler/Mia/its just a test anyway',5),(11,'Write identified proteins by spectral count','yes',6),(12,'Write identified proteins by spot','yes',6),(13,'Write identified proteins by distinct peptide count','yes',6),(14,'Write additional peptides identified by GPF','yes',6),(15,'Write ambiguous peptides','yes',6),(16,'Write modified peptides','yes',6),(17,'Putative entry prefix','__putative__',6),(18,'Output directory','X:/AK Hippler/Mia/its just a test anyway',6),(19,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',6),(20,'HTML report','yes',6),(21,'Write identified proteins by spectral count','yes',7),(22,'Write identified proteins by spot','yes',7),(23,'Write identified proteins by distinct peptide count','yes',7),(24,'Write additional peptides identified by GPF','yes',7),(25,'Write ambiguous peptides','yes',7),(26,'Write modified peptides','yes',7),(27,'Putative entry prefix','__putative__',7),(28,'Output directory','X:/AK Hippler/Mia/its just a test anyway',7),(29,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',7),(30,'HTML report','yes',7),(31,'Write identified proteins by spectral count','yes',8),(32,'Write identified proteins by spot','yes',8),(33,'Write identified proteins by distinct peptide count','yes',8),(34,'Write additional peptides identified by GPF','yes',8),(35,'Write ambiguous peptides','yes',8),(36,'Write modified peptides','yes',8),(37,'Putative entry prefix','__putative__',8),(38,'Output directory','X:/AK Hippler/Mia/its just a test anyway',8),(39,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',8),(40,'HTML report','yes',8),(41,'Write identified proteins by spectral count','yes',9),(42,'Write identified proteins by spot','yes',9),(43,'Write identified proteins by distinct peptide count','yes',9),(44,'Write additional peptides identified by GPF','yes',9),(45,'Write ambiguous peptides','yes',9),(46,'Write modified peptides','yes',9),(47,'Putative entry prefix','__putative__',9),(48,'Output directory','X:/AK Hippler/Mia/its just a test anyway',9),(49,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',9),(50,'HTML report','yes',9),(51,'Write identified proteins by spectral count','yes',10),(52,'Write identified proteins by spot','yes',10),(53,'Write identified proteins by distinct peptide count','yes',10),(54,'Write additional peptides identified by GPF','yes',10),(55,'Write ambiguous peptides','yes',10),(56,'Write modified peptides','yes',10),(57,'Putative entry prefix','__putative__',10),(58,'Output directory','X:/AK Hippler/Mia/its just a test anyway',10),(59,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',10),(60,'HTML report','yes',10),(61,'Write identified proteins by spectral count','yes',11),(62,'Write identified proteins by spot','yes',11),(63,'Write identified proteins by distinct peptide count','yes',11),(64,'Write additional peptides identified by GPF','yes',11),(65,'Write ambiguous peptides','yes',11),(66,'Write modified peptides','yes',11),(67,'Putative entry prefix','__putative__',11),(68,'Output directory','X:/AK Hippler/Mia/its just a test anyway',11),(69,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',11),(70,'HTML report','yes',11),(71,'Write identified proteins by spectral count','yes',12),(72,'Write identified proteins by spot','yes',12),(73,'Write identified proteins by distinct peptide count','yes',12),(74,'Write additional peptides identified by GPF','yes',12),(75,'Write ambiguous peptides','yes',12),(76,'Write modified peptides','yes',12),(77,'Putative entry prefix','__putative__',12),(78,'Output directory','X:/AK Hippler/Mia/its just a test anyway',12),(79,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',12),(80,'HTML report','yes',12),(81,'Write identified proteins by spectral count','yes',13),(82,'Write identified proteins by spot','yes',13),(83,'Write identified proteins by distinct peptide count','yes',13),(84,'Write additional peptides identified by GPF','yes',13),(85,'Write ambiguous peptides','yes',13),(86,'Write modified peptides','yes',13),(87,'Putative entry prefix','__putative__',13),(88,'Output directory','X:/AK Hippler/Mia/its just a test anyway',13),(89,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',13),(90,'HTML report','yes',13),(91,'Write identified proteins by spectral count','yes',14),(92,'Write identified proteins by spot','yes',14),(93,'Write identified proteins by distinct peptide count','yes',14),(94,'Write additional peptides identified by GPF','yes',14),(95,'Write ambiguous peptides','yes',14),(96,'Write modified peptides','yes',14),(97,'Putative entry prefix','__putative__',14),(98,'Output directory','X:/AK Hippler/Mia/its just a test anyway',14),(99,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',14),(100,'HTML report','yes',14),(101,'Write identified proteins by spectral count','yes',15),(102,'Write identified proteins by spot','yes',15),(103,'Write identified proteins by distinct peptide count','yes',15),(104,'Write additional peptides identified by GPF','yes',15),(105,'Write ambiguous peptides','yes',15),(106,'Write modified peptides','yes',15),(107,'Putative entry prefix','__putative__',15),(108,'Output directory','X:/AK Hippler/Mia/its just a test anyway',15),(109,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',15),(110,'HTML report','yes',15),(111,'Write identified proteins by spectral count','yes',16),(112,'Write identified proteins by spot','yes',16),(113,'Write identified proteins by distinct peptide count','yes',16),(114,'Write additional peptides identified by GPF','yes',16),(115,'Write ambiguous peptides','yes',16),(116,'Write modified peptides','yes',16),(117,'Putative entry prefix','__putative__',16),(118,'Output directory','X:/AK Hippler/Mia/its just a test anyway',16),(119,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',16),(120,'HTML report','yes',16),(121,'Write identified proteins by spectral count','yes',17),(122,'Write identified proteins by spot','yes',17),(123,'Write identified proteins by distinct peptide count','yes',17),(124,'Write additional peptides identified by GPF','yes',17),(125,'Write ambiguous peptides','yes',17),(126,'Write modified peptides','yes',17),(127,'Putative entry prefix','__putative__',17),(128,'Output directory','X:/AK Hippler/Mia/its just a test anyway',17),(129,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',17),(130,'HTML report','yes',17),(131,'Write identified proteins by spectral count','yes',18),(132,'Write identified proteins by spot','yes',18),(133,'Write identified proteins by distinct peptide count','yes',18),(134,'Write additional peptides identified by GPF','yes',18),(135,'Write ambiguous peptides','yes',18),(136,'Write modified peptides','yes',18),(137,'Putative entry prefix','__putative__',18),(138,'Output directory','X:/AK Hippler/Mia/its just a test anyway',18),(139,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',18),(140,'HTML report','yes',18),(141,'Write identified proteins by spectral count','yes',19),(142,'Write identified proteins by spot','yes',19),(143,'Write identified proteins by distinct peptide count','yes',19),(144,'Write additional peptides identified by GPF','yes',19),(145,'Write ambiguous peptides','yes',19),(146,'Write modified peptides','yes',19),(147,'Putative entry prefix','__putative__',19),(148,'Output directory','X:/AK Hippler/Mia/its just a test anyway',19),(149,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',19),(150,'HTML report','yes',19),(151,'Write identified proteins by spectral count','yes',20),(152,'Write identified proteins by spot','yes',20),(153,'Write identified proteins by distinct peptide count','yes',20),(154,'Write additional peptides identified by GPF','yes',20),(155,'Write ambiguous peptides','yes',20),(156,'Write modified peptides','yes',20),(157,'Putative entry prefix','__putative__',20),(158,'Output directory','X:/AK Hippler/Mia/its just a test anyway',20),(159,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',20),(160,'HTML report','yes',20),(161,'Write identified proteins by spectral count','yes',21),(162,'Write identified proteins by spot','yes',21),(163,'Write identified proteins by distinct peptide count','yes',21),(164,'Write additional peptides identified by GPF','yes',21),(165,'Write ambiguous peptides','yes',21),(166,'Write modified peptides','yes',21),(167,'Putative entry prefix','__putative__',21),(168,'Output directory','X:/AK Hippler/Mia/its just a test anyway',21),(169,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',21),(170,'HTML report','yes',21),(171,'Write identified proteins by spectral count','yes',22),(172,'Write identified proteins by spot','yes',22),(173,'Write identified proteins by distinct peptide count','yes',22),(174,'Write additional peptides identified by GPF','yes',22),(175,'Write ambiguous peptides','yes',22),(176,'Write modified peptides','yes',22),(177,'Putative entry prefix','__putative__',22),(178,'Output directory','X:/AK Hippler/Mia/its just a test anyway',22),(179,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',22),(180,'HTML report','yes',22),(181,'Write identified proteins by spectral count','yes',23),(182,'Write identified proteins by spot','yes',23),(183,'Write identified proteins by distinct peptide count','yes',23),(184,'Write additional peptides identified by GPF','yes',23),(185,'Write ambiguous peptides','yes',23),(186,'Write modified peptides','yes',23),(187,'Putative entry prefix','__putative__',23),(188,'Output directory','X:/AK Hippler/Mia/its just a test anyway',23),(189,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',23),(190,'HTML report','yes',23),(191,'Write identified proteins by spectral count','yes',24),(192,'Write identified proteins by spot','yes',24),(193,'Write identified proteins by distinct peptide count','yes',24),(194,'Write additional peptides identified by GPF','yes',24),(195,'Write ambiguous peptides','yes',24),(196,'Write modified peptides','yes',24),(197,'Putative entry prefix','__putative__',24),(198,'Output directory','X:/AK Hippler/Mia/its just a test anyway',24),(199,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',24),(200,'HTML report','yes',24),(201,'Write identified proteins by spectral count','yes',25),(202,'Write identified proteins by spot','yes',25),(203,'Write identified proteins by distinct peptide count','yes',25),(204,'Write additional peptides identified by GPF','yes',25),(205,'Write ambiguous peptides','yes',25),(206,'Write modified peptides','yes',25),(207,'Putative entry prefix','__putative__',25),(208,'Output directory','X:/AK Hippler/Mia/its just a test anyway',25),(209,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',25),(210,'HTML report','yes',25),(211,'Write identified proteins by spectral count','yes',26),(212,'Write identified proteins by spot','yes',26),(213,'Write identified proteins by distinct peptide count','yes',26),(214,'Write additional peptides identified by GPF','yes',26),(215,'Write ambiguous peptides','yes',26),(216,'Write modified peptides','yes',26),(217,'Putative entry prefix','__putative__',26),(218,'Output directory','X:/AK Hippler/Mia/its just a test anyway',26),(219,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',26),(220,'HTML report','yes',26),(221,'Write identified proteins by spectral count','yes',27),(222,'Write identified proteins by spot','yes',27),(223,'Write identified proteins by distinct peptide count','yes',27),(224,'Write additional peptides identified by GPF','yes',27),(225,'Write ambiguous peptides','yes',27),(226,'Write modified peptides','yes',27),(227,'Putative entry prefix','__putative__',27),(228,'Output directory','X:/AK Hippler/Mia/its just a test anyway',27),(229,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',27),(230,'HTML report','yes',27),(231,'Write identified proteins by spectral count','yes',28),(232,'Write identified proteins by spot','yes',28),(233,'Write identified proteins by distinct peptide count','yes',28),(234,'Write additional peptides identified by GPF','yes',28),(235,'Write ambiguous peptides','yes',28),(236,'Write modified peptides','yes',28),(237,'Putative entry prefix','__putative__',28),(238,'Output directory','X:/AK Hippler/Mia/its just a test anyway',28),(239,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',28),(240,'HTML report','yes',28),(241,'Write identified proteins by spectral count','yes',29),(242,'Write identified proteins by spot','yes',29),(243,'Write identified proteins by distinct peptide count','yes',29),(244,'Write additional peptides identified by GPF','yes',29),(245,'Write ambiguous peptides','yes',29),(246,'Write modified peptides','yes',29),(247,'Putative entry prefix','__putative__',29),(248,'Output directory','X:/AK Hippler/Mia/its just a test anyway',29),(249,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',29),(250,'HTML report','yes',29),(251,'Write identified proteins by spectral count','yes',30),(252,'Write identified proteins by spot','yes',30),(253,'Write identified proteins by distinct peptide count','yes',30),(254,'Write additional peptides identified by GPF','yes',30),(255,'Write ambiguous peptides','yes',30),(256,'Write modified peptides','yes',30),(257,'Putative entry prefix','__putative__',30),(258,'Output directory','X:/AK Hippler/Mia/its just a test anyway',30),(259,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',30),(260,'HTML report','yes',30),(261,'Write identified proteins by spectral count','yes',31),(262,'Write identified proteins by spot','yes',31),(263,'Write identified proteins by distinct peptide count','yes',31),(264,'Write additional peptides identified by GPF','yes',31),(265,'Write ambiguous peptides','yes',31),(266,'Write modified peptides','yes',31),(267,'Putative entry prefix','__putative__',31),(268,'Output directory','X:/AK Hippler/Mia/its just a test anyway',31),(269,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',31),(270,'HTML report','yes',31),(271,'Write identified proteins by spectral count','yes',32),(272,'Write identified proteins by spot','yes',32),(273,'Write identified proteins by distinct peptide count','yes',32),(274,'Write additional peptides identified by GPF','yes',32),(275,'Write ambiguous peptides','yes',32),(276,'Write modified peptides','yes',32),(277,'Putative entry prefix','__putative__',32),(278,'Output directory','X:/AK Hippler/Mia/its just a test anyway',32),(279,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',32),(280,'HTML report','yes',32),(281,'Write identified proteins by spectral count','yes',33),(282,'Write identified proteins by spot','yes',33),(283,'Write identified proteins by distinct peptide count','yes',33),(284,'Write additional peptides identified by GPF','yes',33),(285,'Write ambiguous peptides','yes',33),(286,'Write modified peptides','yes',33),(287,'Putative entry prefix','__putative__',33),(288,'Output directory','X:/AK Hippler/Mia/its just a test anyway',33),(289,'Output file prefix','MT_whole cells-omssa-results-filtered-by-fpr-filtered-by-mass-accuracy-',33),(290,'HTML report','yes',33),(291,'Write identified proteins by spectral count','yes',34),(292,'Write identified proteins by spot','yes',34),(293,'Write identified proteins by distinct peptide count','yes',34),(294,'Write additional peptides identified by GPF','yes',34),(295,'Write ambiguous peptides','yes',34),(296,'Write modified peptides','yes',34),(297,'Output directory','',34),(298,'Output file prefix','Grill_Sample1_all_omssa-results-1Prozentglobalstrict_filtered-by-fpr-filtered-by-mass-accuracy-',34),(299,'HTML report','yes',34),(300,'Write identified proteins by spectral count','yes',35),(301,'Write identified proteins by spot','yes',35),(302,'Write identified proteins by distinct peptide count','yes',35),(303,'Write additional peptides identified by GPF','yes',35),(304,'Write ambiguous peptides','yes',35),(305,'Write modified peptides','yes',35),(306,'Output directory','',35),(307,'Output file prefix','Grill_Sample1_all_omssa-results-1Prozentglobalstrict_filtered-by-fpr-filtered-by-mass-accuracy-',35),(308,'HTML report','yes',35),(309,'Write identified proteins by spectral count','yes',36),(310,'Write identified proteins by spot','yes',36),(311,'Write identified proteins by distinct peptide count','yes',36),(312,'Write additional peptides identified by GPF','yes',36),(313,'Write ambiguous peptides','yes',36),(314,'Write modified peptides','yes',36),(315,'Output directory','',36),(316,'Output file prefix','Grill_Sample1_all_omssa-results-1Prozentglobalstrict_filtered-by-fpr-filtered-by-mass-accuracy-',36),(317,'HTML report','yes',36),(318,'Write identified proteins by spectral count','yes',37),(319,'Write identified proteins by spot','yes',37),(320,'Write identified proteins by distinct peptide count','yes',37),(321,'Write additional peptides identified by GPF','yes',37),(322,'Write ambiguous peptides','yes',37),(323,'Write modified peptides','yes',37),(324,'Output directory','',37),(325,'Output file prefix','Grill_Sample1_all_omssa-results-1Prozentglobalstrict_filtered-by-fpr-filtered-by-mass-accuracy-',37),(326,'HTML report','yes',37),(327,'Precursor ion m/z tolerance','0.02 Da',38),(328,'Product ion m/z tolerance','0.5 Da',38),(329,'Precursor ion search type','monoisotopic',38),(330,'Product ion search type','monoisotopic',38),(331,'Precursor charge dependency','none',38),(332,'Neutron mass add threshold','1446.94 Da',38),(333,'Low intensity cutoff','0.0',38),(334,'Single charge window','20 Da',38),(335,'Double charge window','14 Da',38),(336,'Min peaks in single charge window','2',38),(337,'Min peaks in double charge window','2',38),(338,'Eliminate charge reduced precursors','no',38),(339,'Min peak count (spectrum)','4',38),(340,'Min precursor spectrum match count','1',38),(341,'Proline exception','-',38),(342,'Min precursor ion charge','1',38),(343,'Max precursor ion charge','4',38),(344,'Min charge for considering multi','3',38),(345,'Peaks fraction below precursor for +1','0.95',38),(346,'Determine charge +1','yes',38),(347,'Precursor charge determination','use a range',38),(348,'Max product ion charge','3',38),(349,'Missed cleavages allowed','2',38),(350,'Enzyme','Trypsin',38),(351,'Min peptide length for NE and ST searches','4',38),(352,'Max peptide length for NE and ST searches','40',38),(353,'Ions to search','b, y',38),(354,'Search c-term ions','yes',38),(355,'Max ions to search per series','100',38),(356,'Search first forward b1 product ions','no',38),(357,'Taxonomy IDs to search','0',38),(358,'Cleave n-term methionine','yes',38),(359,'Fixed modifications','-',38),(360,'Variable modifications','phosphorylation of S, phosphorylation of T, phosphorylation of Y',38),(361,'Min matching peaks (database)','2',38),(362,'Min matching peaks (spectrum)','6',38),(363,'Use memory mapped sequence libraries','no',38),(364,'Maximum hit list size','30',38),(365,'Mass ladders per database peptide','128',38),(366,'Use correlation correction','yes',38),(367,'Auto mass tolerance adjustment','1.0',38),(368,'Consecutive ion probability','0.5',38),(369,'e value cutoff','1.0',38),(370,'Report spectra and search parameters','no',38),(371,'Perform target-decoy search','no',38),(372,'Target-decoy method','shuffle peptide',38),(373,'Decoy keep start','0',38),(374,'Decoy keep end','1',38),(375,'Spectrum batch size','2000',38),(376,'Output directory','/home/michael/ingrid/omssa/sixframes/WT+CaWCAN__Mix_211108',38),(377,'Output file prefix','IH_WT+CaAN_Mix_211108_20-',38),(378,'OMSSA results','yes',38),(379,'Precursor ion m/z tolerance','0.02 Da',39),(380,'Product ion m/z tolerance','0.5 Da',39),(381,'Precursor ion search type','monoisotopic',39),(382,'Product ion search type','monoisotopic',39),(383,'Precursor charge dependency','none',39),(384,'Neutron mass add threshold','1446.94 Da',39),(385,'Low intensity cutoff','0.0',39),(386,'Single charge window','20 Da',39),(387,'Double charge window','14 Da',39),(388,'Min peaks in single charge window','2',39),(389,'Min peaks in double charge window','2',39),(390,'Eliminate charge reduced precursors','no',39),(391,'Min peak count (spectrum)','4',39),(392,'Min precursor spectrum match count','1',39),(393,'Proline exception','-',39),(394,'Min precursor ion charge','1',39),(395,'Max precursor ion charge','4',39),(396,'Min charge for considering multi','3',39),(397,'Peaks fraction below precursor for +1','0.95',39),(398,'Determine charge +1','yes',39),(399,'Precursor charge determination','use a range',39),(400,'Max product ion charge','3',39),(401,'Missed cleavages allowed','2',39),(402,'Enzyme','Trypsin',39),(403,'Min peptide length for NE and ST searches','4',39),(404,'Max peptide length for NE and ST searches','40',39),(405,'Ions to search','b, y',39),(406,'Search c-term ions','yes',39),(407,'Max ions to search per series','100',39),(408,'Search first forward b1 product ions','no',39),(409,'Taxonomy IDs to search','0',39),(410,'Cleave n-term methionine','yes',39),(411,'Fixed modifications','-',39),(412,'Variable modifications','phosphorylation of S, phosphorylation of T, phosphorylation of Y',39),(413,'Min matching peaks (database)','2',39),(414,'Min matching peaks (spectrum)','6',39),(415,'Use memory mapped sequence libraries','no',39),(416,'Maximum hit list size','30',39),(417,'Mass ladders per database peptide','128',39),(418,'Use correlation correction','yes',39),(419,'Auto mass tolerance adjustment','1.0',39),(420,'Consecutive ion probability','0.5',39),(421,'e value cutoff','1.0',39),(422,'Report spectra and search parameters','no',39),(423,'Perform target-decoy search','no',39),(424,'Target-decoy method','shuffle peptide',39),(425,'Decoy keep start','0',39),(426,'Decoy keep end','1',39),(427,'Spectrum batch size','2000',39),(428,'Output directory','/home/michael/ingrid/omssa/sixframes/WT+CaWCAN__Mix_211108',39),(429,'Output file prefix','IH_WT+CaAN_Mix_211108_20-',39),(430,'OMSSA results','yes',39),(431,'Precursor ion m/z tolerance','0.02 Da',40),(432,'Product ion m/z tolerance','0.5 Da',40),(433,'Precursor ion search type','monoisotopic',40),(434,'Product ion search type','monoisotopic',40),(435,'Precursor charge dependency','none',40),(436,'Neutron mass add threshold','1446.94 Da',40),(437,'Low intensity cutoff','0.0',40),(438,'Single charge window','20 Da',40),(439,'Double charge window','14 Da',40),(440,'Min peaks in single charge window','2',40),(441,'Min peaks in double charge window','2',40),(442,'Eliminate charge reduced precursors','no',40),(443,'Min peak count (spectrum)','4',40),(444,'Min precursor spectrum match count','1',40),(445,'Proline exception','-',40),(446,'Min precursor ion charge','1',40),(447,'Max precursor ion charge','4',40),(448,'Min charge for considering multi','3',40),(449,'Peaks fraction below precursor for +1','0.95',40),(450,'Determine charge +1','yes',40),(451,'Precursor charge determination','use a range',40),(452,'Max product ion charge','3',40),(453,'Missed cleavages allowed','2',40),(454,'Enzyme','Trypsin',40),(455,'Min peptide length for NE and ST searches','4',40),(456,'Max peptide length for NE and ST searches','40',40),(457,'Ions to search','b, y',40),(458,'Search c-term ions','yes',40),(459,'Max ions to search per series','100',40),(460,'Search first forward b1 product ions','no',40),(461,'Taxonomy IDs to search','0',40),(462,'Cleave n-term methionine','yes',40),(463,'Fixed modifications','-',40),(464,'Variable modifications','phosphorylation of S, phosphorylation of T, phosphorylation of Y',40),(465,'Min matching peaks (database)','2',40),(466,'Min matching peaks (spectrum)','6',40),(467,'Use memory mapped sequence libraries','no',40),(468,'Maximum hit list size','30',40),(469,'Mass ladders per database peptide','128',40),(470,'Use correlation correction','yes',40),(471,'Auto mass tolerance adjustment','1.0',40),(472,'Consecutive ion probability','0.5',40),(473,'e value cutoff','1.0',40),(474,'Report spectra and search parameters','no',40),(475,'Perform target-decoy search','no',40),(476,'Target-decoy method','shuffle peptide',40),(477,'Decoy keep start','0',40),(478,'Decoy keep end','1',40),(479,'Spectrum batch size','2000',40),(480,'Output directory','/home/michael/ingrid/omssa/sixframes/WT+CaWCAN__Mix_211108',40),(481,'Output file prefix','IH_WT+CaAN_Mix_211108_20-',40),(482,'OMSSA results','yes',40),(483,'Masses','monoisotopic',41),(484,'Similarity search','yes',41),(485,'Mass tolerance','700 ppm',41),(486,'Search introns','yes',41),(487,'Max intron length','2100 nucl.',41),(488,'Enzyme','Trypsin',41),(489,'Minimum amino acid chain length','5',41),(490,'Output directory','',41),(491,'Output file prefix','MT_CPM_AN_030409_31-',41),(492,'GPF peptides','yes',41),(493,'GPF results','no',41);
/*!40000 ALTER TABLE `parameters` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `run_filewithname`
--

DROP TABLE IF EXISTS `run_filewithname`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `run_filewithname` (
  `run_id` int(10) unsigned NOT NULL,
  `filewithname_id` int(10) unsigned NOT NULL,
  `input_file` tinyint(1) NOT NULL,
  KEY `run_id` (`run_id`),
  KEY `filewithname` (`filewithname_id`) USING BTREE,
  CONSTRAINT `filewithname_id` FOREIGN KEY (`filewithname_id`) REFERENCES `filewithname` (`filewithname_id`),
  CONSTRAINT `run_id2` FOREIGN KEY (`run_id`) REFERENCES `runs` (`run_id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `run_filewithname`
--

LOCK TABLES `run_filewithname` WRITE;
/*!40000 ALTER TABLE `run_filewithname` DISABLE KEYS */;
/*!40000 ALTER TABLE `run_filewithname` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `runs`
--

DROP TABLE IF EXISTS `runs`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `runs` (
  `run_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `user` varchar(1024) NOT NULL,
  `title` varchar(1024) NOT NULL,
  `host` varchar(1024) NOT NULL,
  `script_uri` varchar(1024) NOT NULL,
  `version` varchar(1024) NOT NULL,
  `start_time` datetime NOT NULL,
  `end_time` datetime NOT NULL,
  PRIMARY KEY (`run_id`)
) ENGINE=InnoDB AUTO_INCREMENT=42 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `runs`
--

LOCK TABLES `runs` WRITE;
/*!40000 ALTER TABLE `runs` DISABLE KEYS */;
INSERT INTO `runs` VALUES (1,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(2,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(3,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(4,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(5,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(6,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(7,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(8,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(9,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(10,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(11,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(12,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(13,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(14,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(15,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(16,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(17,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(18,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(19,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(20,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(21,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(22,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(23,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(24,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(25,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(26,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(27,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(28,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(29,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(30,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(31,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(32,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(33,'mtera_01','Write HTML report','IBBP-110-159-1','write-omssa-report.rb','trunk','2009-06-10 11:35:33','2009-06-10 11:35:37'),(34,'r0bp_admin','Write HTML report','IBBP-110-167-2','write-omssa-report.rb','trunk','2009-05-25 15:56:00','2009-05-25 15:56:11'),(35,'r0bp_admin','Write HTML report','IBBP-110-167-2','write-omssa-report.rb','trunk','2009-05-25 15:56:00','2009-05-25 15:56:11'),(36,'r0bp_admin','Write HTML report','IBBP-110-167-2','write-omssa-report.rb','trunk','2009-05-25 15:56:00','2009-05-25 15:56:11'),(37,'r0bp_admin','Write HTML report','IBBP-110-167-2','write-omssa-report.rb','trunk','2009-05-25 15:56:00','2009-05-25 15:56:11'),(38,'michael','Run OMSSA','peaks','run-omssa.rb','trunk','2009-05-15 06:58:53','2009-05-15 07:20:28'),(39,'michael','Run OMSSA','peaks','run-omssa.rb','trunk','2009-05-15 06:58:53','2009-05-15 07:20:28'),(40,'michael','Run OMSSA','peaks','run-omssa.rb','trunk','2009-05-15 06:58:53','2009-05-15 07:20:28'),(41,'michael','Run GPF','gpf','run-gpf.rb','trunk','2009-06-05 22:47:21','2009-06-05 23:26:03');
/*!40000 ALTER TABLE `runs` ENABLE KEYS */;
UNLOCK TABLES;
/*!40103 SET TIME_ZONE=@OLD_TIME_ZONE */;

/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40014 SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;

-- Dump completed on 2009-07-06  9:48:59
